core/sdk-core/src/main/java/software/amazon/awssdk/core/client/builder/SdkDefaultClientBuilder.java

```diff
@@ -315,9 +320,59 @@ private SdkClientConfiguration finalizeConfiguration(SdkClientConfiguration conf
                      .option(EXECUTION_INTERCEPTORS, resolveExecutionInterceptors(config))
                      .option(RETRY_POLICY, retryPolicy)
                      .option(CLIENT_USER_AGENT, resolveClientUserAgent(config, retryPolicy))
+                     .option(REQUEST_COMPRESSION_CONFIGURATION, resolveRequestCompressionConfiguration())
                      .build();
     }
 
+    private RequestCompressionConfiguration resolveRequestCompressionConfiguration() {
```

Comment:
Configuration in SdkDefaultClientBuilder seems confusing.

What I'd expect to see:
1. In `setOverrides`, you set the REQUEST_COMPRESSION_CONFIGURATION to what was in the client override configuration.
2. In `mergeGlobalDefaults`, if the REQUEST_COMPRESSION_CONFIGURATION is null, check the environment variables/system properties. If those aren't set, use the PROFILE_FILE_SUPPLIER and PROFILE_NAME to check the profile file (they will never be null there).

You might also be able to use Optional and OptionalUtils.firstPresent to remove a lot of these null checks.

Reply:
refactored to update in `setOverrides()` and `mergeGlobalDefaults()`

added new method `addCompressionConfigGlobalDefaults()` to handle global defaults.
even if COMPRESSION_CONFIGURATION is not null, we still need to check both fields in the config object since one of them could be null